InvenTree/plugin/base/barcodes/mixins.py

```diff
@@ -37,3 +51,203 @@ def scan(self, barcode_data):
+                    barcode_fields[field_name] = entry[len(identifier):]
+                    break
+
+        return barcode_fields
+
+    @staticmethod
+    def parse_isoiec_15434_barcode2d(barcode_data: str) -> list[str]:
+        """Parse a ISO/IEC 15434 bardode, returning the split data section."""
+        HEADER = "[)>\x1E06\x1D"
+        TRAILER = "\x1E\x04"
+
+        # some old mouser barcodes start with this messed up header
+        OLD_MOUSER_HEADER = ">[)>06\x1D"
+        if barcode_data.startswith(OLD_MOUSER_HEADER):
+            barcode_data = barcode_data.replace(OLD_MOUSER_HEADER, HEADER, 1)
+
+        # most barcodes don't include the trailer, because "why would you stick to
+        # the standard, right?" so we only check for the header here
+        if not barcode_data.startswith(HEADER):
+            return
+
+        actual_data = barcode_data.split(HEADER, 1)[1].rsplit(TRAILER, 1)[0]
+
+        return actual_data.split("\x1D")
+
+    @staticmethod
+    def receive_purchase_order_item(
+            supplier_part: SupplierPart,
+            user: User,
+            quantity: int | str = None,
+            order_number: str = None,
+            purchase_order: PurchaseOrder = None,
+            location: StockLocation = None,
+            barcode: str = None,
+    ) -> dict:
+        """Try to receive a purchase order item.
+
+        Returns a dict object containing:
+            - on success: a "success" message
+            - on partial success: the "lineitem" with quantity and location (both can be None)
+            - on failure: an "error" message
+        """
+
+        if not purchase_order:
+            # try to find a purchase order with either reference or name matching
+            # the provided order_number
+            if not order_number:
+                return {"error": _("Supplier barcode doesn't contain order number")}
+
+            purchase_orders = (
+                PurchaseOrder.objects.filter(
+                    supplier_reference__iexact=order_number,
+                    status=PurchaseOrderStatus.PLACED.value,
+                ) | PurchaseOrder.objects.filter(
+                    reference__iexact=order_number,
+                    status=PurchaseOrderStatus.PLACED.value,
+                )
+            )
+
+            if len(purchase_orders) == 0:
+                return {"error": _(f"Failed to find placed purchase order for '{order_number}'")}
+            elif len(purchase_orders) > 1:
+                return {"error": _(f"Found multiple placed purchase orders for '{order_number}'")}
+
+            purchase_order = purchase_orders[0]
+
+        #  find the first incomplete line_item that matches the supplier_part
+        line_items = purchase_order.lines.filter(
+            part=supplier_part.pk, quantity__gt=F("received"))
+        if not line_items:
+            return {"error": _("Failed to find pending line item for supplier part")}
+
```

Comment:
Well as the comment above says, my idea was pretty much that it'd just be fine to return the **first** one then. I think duplicated SKUs in the same purchase order would already be an edge case, right? Not sure how that would happen really.
But anyways, so in those cases, once you confirm enough "quantity" of the first line item, this would automatically switch to the next one.

Reply:
Having the same SKU multiple times is easy if you order for multiple BOs/multiple projects in one PO.

Reply:
Ah okay interesting, well one possible improvement to this I could see would be matching the quantity, if supplied aswell.
I.e. if we have multiple line items, check each of them and see if one of them matches the quantity provided in the barcode. If not, just match the first one (as it's now).

Reply:
@30350n that is almost sufficient, I think. You will also need to ensure that the received quantity makes sense for the "first" item, if there is no exact match. What if you receive a partial quantity shipment?

Reply:
That shouldn't be a problem. In those cases a stock item for the partial quantity gets created, but the line item is still "unfinished". So once you scan the item containing "the rest", it returns the same line item again.

> You will also need to ensure that the received quantity makes sense for the "first" item, if there is no exact match.

That would mean the line items quantity should ideally be greater than the barcode quantity in those cases then, right?

Reply:
Yes, that's right.

Reply:
Should be good now (I added a bunch of for else logic :P)